File: src/makoto/dsse.py
"""DSSE PAE plus the strict Makoto v0.2 Ed25519 profile."""

from __future__ import annotations

import base64
import binascii
import hashlib
import re
from collections.abc import Sequence
from dataclasses import dataclass

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey

from makoto.digest import sha256_bytes

P = 2**255 - 19
L = 2**252 + 27742317777372353535851937790883648493
D = (-121665 * pow(121666, P - 2, P)) % P
SQRT_MINUS_ONE = pow(2, (P - 1) // 4, P)
IDENTITY = (0, 1)
BASE_Y = (4 * pow(5, P - 2, P)) % P
_BASE_X_SQUARED = ((BASE_Y * BASE_Y - 1) * pow(D * BASE_Y * BASE_Y + 1, P - 2, P)) % P
_BASE_X = pow(_BASE_X_SQUARED, (P + 3) // 8, P)
if (_BASE_X * _BASE_X - _BASE_X_SQUARED) % P != 0:
    _BASE_X = (_BASE_X * SQRT_MINUS_ONE) % P
if _BASE_X & 1:
    _BASE_X = P - _BASE_X
BASE = (_BASE_X, BASE_Y)
SPKI_PREFIX = bytes.fromhex("302a300506032b6570032100")
SUPPORTED_PAYLOAD_TYPES = frozenset(
    {
        "application/vnd.in-toto+json",
        "application/vnd.makoto.handoff.v0.2+json",
    }
)
_PAYLOAD_TYPE = re.compile(r"^[a-z0-9][a-z0-9!#$&^_.+\-]{0,126}/[a-z0-9][a-z0-9!#$&^_.+\-]{0,126}$")


class DsseError(ValueError):
    """Raised when DSSE or key material is malformed."""


class SignatureVerificationError(DsseError):
    """Raised when a strict Ed25519 verification fails."""


@dataclass(frozen=True)
class SigningKey:
    private_key: Ed25519PrivateKey

    @classmethod
    def generate(cls) -> SigningKey:
        return cls(Ed25519PrivateKey.generate())

    @classmethod
    def from_seed(cls, seed: bytes) -> SigningKey:
        if len(seed) != 32:
            raise DsseError("Ed25519 seed must be exactly 32 bytes")
        return cls(Ed25519PrivateKey.from_private_bytes(seed))

    @classmethod
    def from_pem(cls, data: bytes) -> SigningKey:
        try:
            key = serialization.load_pem_private_key(data, password=None)
        except (TypeError, ValueError) as error:
            raise DsseError("private key is not unencrypted PKCS#8 PEM") from error
        if not isinstance(key, Ed25519PrivateKey):
            raise DsseError("private key is not Ed25519")
        return cls(key)

    def public_key_bytes(self) -> bytes:
        return self.private_key.public_key().public_bytes(
            encoding=serialization.Encoding.Raw,
            format=serialization.PublicFormat.Raw,
        )

    def public_spki(self) -> bytes:
        return SPKI_PREFIX + self.public_key_bytes()

    def private_pkcs8_pem(self) -> bytes:
        return self.private_key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.PKCS8,
            encryption_algorithm=serialization.NoEncryption(),
        )

    def public_spki_pem(self) -> bytes:
        return self.private_key.public_key().public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo,
        )

    def keyid(self) -> str:
        return keyid_from_spki(self.public_spki())

    def sign(self, message: bytes) -> bytes:
        return self.private_key.sign(message)


def pae(payload_type: str, payload: bytes) -> bytes:
    try:
        payload_type_bytes = payload_type.encode("utf-8", errors="strict")
    except UnicodeEncodeError as error:
        raise DsseError("payload type is not Unicode scalar text") from error
    return b" ".join(
        (
            b"DSSEv1",
            str(len(payload_type_bytes)).encode("ascii"),
            payload_type_bytes,
            str(len(payload)).encode("ascii"),
            payload,
        )
    )


def validate_payload_type(payload_type: str, *, require_supported: bool = True) -> None:
    """Validate the exact Makoto v0.2 DSSE payload-type grammar and allowlist."""

    if _PAYLOAD_TYPE.fullmatch(payload_type) is None:
        raise DsseError("payload type is not a lowercase parameter-free media type")
    if require_supported and payload_type not in SUPPORTED_PAYLOAD_TYPES:
        raise DsseError("payload type is not supported by Makoto v0.2")


def canonical_b64encode(data: bytes) -> str:
    return base64.b64encode(data).decode("ascii")


def canonical_b64decode(value: str, *, expected_length: int | None = None) -> bytes:
    try:
        ascii_value = value.encode("ascii", errors="strict")
        decoded = base64.b64decode(ascii_value, validate=True)
    except (UnicodeEncodeError, binascii.Error, ValueError) as error:
        raise DsseError("value is not canonical RFC 4648 standard base64") from error
    if canonical_b64encode(decoded) != value:
        raise DsseError("value is not the canonical base64 spelling")
    if expected_length is not None and len(decoded) != expected_length:
        raise DsseError(f"decoded value must be exactly {expected_length} bytes")
    return decoded


def keyid_from_spki(spki: bytes) -> str:
    if len(spki) != 44 or not spki.startswith(SPKI_PREFIX):
        raise DsseError("public key is not canonical Ed25519 SubjectPublicKeyInfo")
    _decode_prime_order_point(spki[len(SPKI_PREFIX) :])
    return f"sha256:{sha256_bytes(spki)}"


def public_key_from_spki(spki: bytes) -> bytes:
    keyid_from_spki(spki)
    return spki[len(SPKI_PREFIX) :]


def spki_from_pem(data: bytes) -> bytes:
    try:
        key = serialization.load_pem_public_key(data)
    except ValueError as error:
        raise DsseError("public key is not SubjectPublicKeyInfo PEM") from error
    if not isinstance(key, Ed25519PublicKey):
        raise DsseError("public key is not Ed25519")
    spki = key.public_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    keyid_from_spki(spki)
    return spki


def _inverse(value: int) -> int:
    if value % P == 0:
        raise SignatureVerificationError("invalid Edwards25519 denominator")
    return pow(value, P - 2, P)


ExtendedPoint = tuple[int, int, int, int]


def _to_extended(point: tuple[int, int]) -> ExtendedPoint:
    x, y = point
    return x, y, 1, x * y % P


def _extended_add(left: ExtendedPoint, right: ExtendedPoint) -> ExtendedPoint:
    """Complete extended-coordinate addition for Edwards25519 (a = -1)."""

    x1, y1, z1, t1 = left
    x2, y2, z2, t2 = right
    a = (y1 - x1) * (y2 - x2) % P
    b = (y1 + x1) * (y2 + x2) % P
    c = 2 * D * t1 * t2 % P
    d = 2 * z1 * z2 % P
    e = b - a
    f = d - c
    g = d + c
    h = b + a
    return e * f % P, g * h % P, f * g % P, e * h % P


def _from_extended(point: ExtendedPoint) -> tuple[int, int]:
    x, y, z, _t = point
    inverse_z = _inverse(z)
    return x * inverse_z % P, y * inverse_z % P


def _add(left: tuple[int, int], right: tuple[int, int]) -> tuple[int, int]:
    return _from_extended(_extended_add(_to_extended(left), _to_extended(right)))


def _scalar_multiply(scalar: int, point: tuple[int, int]) -> tuple[int, int]:
    result = _to_extended(IDENTITY)
    addend = _to_extended(point)
    while scalar:
        if scalar & 1:
            result = _extended_add(result, addend)
        addend = _extended_add(addend, addend)
        scalar >>= 1
    return _from_extended(result)


def _encode_point(point: tuple[int, int]) -> bytes:
    x, y = point
    encoded = y | ((x & 1) << 255)
    return encoded.to_bytes(32, "little")


def _decode_prime_order_point(encoded: bytes) -> tuple[int, int]:
    if len(encoded) != 32:
        raise SignatureVerificationError("Edwards25519 point must be 32 bytes")
    encoded_integer = int.from_bytes(encoded, "little")
    sign = encoded_integer >> 255
    y = encoded_integer & ((1 << 255) - 1)
    if y >= P:
        raise SignatureVerificationError("noncanonical Edwards25519 y coordinate")
    y_squared = y * y % P
    x_squared = ((y_squared - 1) * _inverse(D * y_squared + 1)) % P
    x = pow(x_squared, (P + 3) // 8, P)
    if (x * x - x_squared) % P != 0:
        x = x * SQRT_MINUS_ONE % P
    if (x * x - x_squared) % P != 0:
        raise SignatureVerificationError("Edwards25519 point is not on the curve")
    if x == 0 and sign == 1:
        raise SignatureVerificationError("invalid sign bit for zero x coordinate")
    if (x & 1) != sign:
        x = P - x
    point = (x, y)
    if _encode_point(point) != encoded:
        raise SignatureVerificationError("noncanonical Edwards25519 point encoding")
    if point == IDENTITY:
        raise SignatureVerificationError("identity Edwards25519 point is forbidden")
    if _scalar_multiply(L, point) != IDENTITY:
        raise SignatureVerificationError("Edwards25519 point is not prime order")
    return point


def strict_verify_ed25519(public_key: bytes, signature: bytes, message: bytes) -> None:
    if len(public_key) != 32:
        raise SignatureVerificationError("Ed25519 public key must be 32 bytes")
    if len(signature) != 64:
        raise SignatureVerificationError("Ed25519 signature must be 64 bytes")
    public_point = _decode_prime_order_point(public_key)
    encoded_r = signature[:32]
    signature_point = _decode_prime_order_point(encoded_r)
    scalar = int.from_bytes(signature[32:], "little")
    if scalar >= L:
        raise SignatureVerificationError("noncanonical Ed25519 scalar")
    challenge = (
        int.from_bytes(
            hashlib.sha512(encoded_r + public_key + message).digest(),
            "little",
        )
        % L
    )
    left = _scalar_multiply(scalar, BASE)
    right = _add(signature_point, _scalar_multiply(challenge, public_point))
    if left != right:
        raise SignatureVerificationError("Ed25519 signature equation failed")


def sign_envelope(
    payload_type: str,
    payload: bytes,
    key: SigningKey | Sequence[SigningKey],
) -> dict[str, object]:
    """Sign exact payload bytes with one or more distinct keys.

    Signature order is canonical by key ID so producer argument order cannot
    change envelope bytes.
    """

    validate_payload_type(payload_type)
    keys = [key] if isinstance(key, SigningKey) else list(key)
    if not keys:
        raise DsseError("at least one signing key is required")
    key_ids = [item.keyid() for item in keys]
    if len(key_ids) != len(set(key_ids)):
        raise DsseError("signing key IDs must be unique")
    message = pae(payload_type, payload)
    signatures = sorted(
        ({"keyid": item.keyid(), "sig": canonical_b64encode(item.sign(message))} for item in keys),
        key=lambda item: item["keyid"].encode(),
    )
    return {
        "payloadType": payload_type,
        "payload": canonical_b64encode(payload),
        "signatures": signatures,
    }


def verify_envelope_signature(
    envelope: dict[str, object],
    *,
    public_spki: bytes,
    signature_index: int = 0,
) -> None:
    payload_type = envelope.get("payloadType")
    payload_text = envelope.get("payload")
    signatures = envelope.get("signatures")
    if not isinstance(payload_type, str) or not isinstance(payload_text, str):
        raise DsseError("envelope payload fields are malformed")
    validate_payload_type(payload_type)
    if not isinstance(signatures, list) or not 0 <= signature_index < len(signatures):
        raise DsseError("signature index is unavailable")
    payload = canonical_b64decode(payload_text)
    seen_keyids: set[str] = set()
    for candidate in signatures:
        if not isinstance(candidate, dict) or set(candidate) != {"keyid", "sig"}:
            raise DsseError("signature entry is malformed")
        candidate_keyid = candidate.get("keyid")
        candidate_signature = candidate.get("sig")
        if not isinstance(candidate_keyid, str) or not isinstance(candidate_signature, str):
            raise DsseError("signature entry is malformed")
        if re.fullmatch(r"sha256:[0-9a-f]{64}", candidate_keyid) is None:
            raise DsseError("signature key ID is malformed")
        if candidate_keyid in seen_keyids:
            raise DsseError("signature key IDs must be unique")
        seen_keyids.add(candidate_keyid)
        canonical_b64decode(candidate_signature, expected_length=64)
    entry = signatures[signature_index]
    if not isinstance(entry, dict):
        raise DsseError("signature entry is malformed")
    keyid = entry.get("keyid")
    signature_text = entry.get("sig")
    expected_keyid = keyid_from_spki(public_spki)
    if keyid != expected_keyid or not isinstance(signature_text, str):
        raise DsseError("signature key ID does not match the supplied public key")
    signature = canonical_b64decode(signature_text, expected_length=64)
    strict_verify_ed25519(public_key_from_spki(public_spki), signature, pae(payload_type, payload))

File: go/dsse.go
package makoto

import (
	"bytes"
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"regexp"
	"sort"
	"strconv"

	"filippo.io/edwards25519"
)

var payloadTypePattern = regexp.MustCompile(
	`^[a-z0-9][a-z0-9!#$&^_.+\-]{0,126}/[a-z0-9][a-z0-9!#$&^_.+\-]{0,126}$`,
)

var keyIDPattern = regexp.MustCompile(`^sha256:[0-9a-f]{64}$`)

var supportedPayloadTypes = map[string]struct{}{
	StatementPayloadType: {},
	HandoffPayloadType:   {},
}

var ed25519Order = [32]byte{
	0xed, 0xd3, 0xf5, 0x5c, 0x1a, 0x63, 0x12, 0x58,
	0xd6, 0x9c, 0xf7, 0xa2, 0xde, 0xf9, 0xde, 0x14,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x10,
}

// PAE returns the DSSE v1 pre-authentication encoding.
func PAE(payloadType string, payload []byte) []byte {
	return bytes.Join(
		[][]byte{
			[]byte("DSSEv1"),
			[]byte(strconv.Itoa(len([]byte(payloadType)))),
			[]byte(payloadType),
			[]byte(strconv.Itoa(len(payload))),
			payload,
		},
		[]byte(" "),
	)
}

// ValidatePayloadType checks the Makoto v0.2 DSSE media-type grammar and
// allowlist.
func ValidatePayloadType(payloadType string) error {
	if !payloadTypePattern.MatchString(payloadType) {
		return fmt.Errorf("payload type is not a lowercase parameter-free media type")
	}
	if _, ok := supportedPayloadTypes[payloadType]; !ok {
		return fmt.Errorf("payload type is not supported by Makoto v0.2")
	}
	return nil
}

func canonicalBase64Decode(value string, expectedLength int) ([]byte, error) {
	decoded, err := base64.StdEncoding.DecodeString(value)
	if err != nil || base64.StdEncoding.EncodeToString(decoded) != value {
		return nil, fmt.Errorf("value is not canonical RFC 4648 standard base64")
	}
	if expectedLength >= 0 && len(decoded) != expectedLength {
		return nil, fmt.Errorf("decoded value must be exactly %d bytes", expectedLength)
	}
	return decoded, nil
}

// GenerateSigningKey creates a new Ed25519 private key.
func GenerateSigningKey() (ed25519.PrivateKey, error) {
	_, privateKey, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("generate Ed25519 key: %w", err)
	}
	return privateKey, nil
}

// SigningKeyFromSeed creates an Ed25519 private key from a 32-byte seed.
func SigningKeyFromSeed(seed []byte) (ed25519.PrivateKey, error) {
	if len(seed) != ed25519.SeedSize {
		return nil, fmt.Errorf("Ed25519 seed must be exactly 32 bytes")
	}
	return ed25519.NewKeyFromSeed(seed), nil
}

// PublicSPKI returns canonical DER SubjectPublicKeyInfo for an Ed25519 key.
func PublicSPKI(publicKey ed25519.PublicKey) ([]byte, error) {
	if len(publicKey) != ed25519.PublicKeySize {
		return nil, fmt.Errorf("Ed25519 public key must be exactly 32 bytes")
	}
	der, err := x509.MarshalPKIXPublicKey(publicKey)
	if err != nil {
		return nil, fmt.Errorf("marshal Ed25519 SubjectPublicKeyInfo: %w", err)
	}
	return der, nil
}

// ParsePublicSPKI parses canonical DER Ed25519 SubjectPublicKeyInfo.
func ParsePublicSPKI(spki []byte) (ed25519.PublicKey, error) {
	parsed, err := x509.ParsePKIXPublicKey(spki)
	if err != nil {
		return nil, fmt.Errorf("public key is not SubjectPublicKeyInfo DER: %w", err)
	}
	publicKey, ok := parsed.(ed25519.PublicKey)
	if !ok {
		return nil, fmt.Errorf("public key is not Ed25519")
	}
	canonical, err := PublicSPKI(publicKey)
	if err != nil || !bytes.Equal(canonical, spki) {
		return nil, fmt.Errorf("public key is not canonical Ed25519 SubjectPublicKeyInfo")
	}
	if err := validatePrimeOrderPoint(publicKey); err != nil {
		return nil, fmt.Errorf("invalid Ed25519 public key: %w", err)
	}
	return publicKey, nil
}

func validatePrimeOrderPoint(encoded []byte) error {
	point, err := new(edwards25519.Point).SetBytes(encoded)
	if err != nil {
		return fmt.Errorf("point is not on the Edwards25519 curve: %w", err)
	}
	if !bytes.Equal(point.Bytes(), encoded) {
		return fmt.Errorf("point encoding is noncanonical")
	}
	identity := edwards25519.NewIdentityPoint()
	if point.Equal(identity) == 1 {
		return fmt.Errorf("identity Edwards25519 point is forbidden")
	}
	result := edwards25519.NewIdentityPoint()
	addend := new(edwards25519.Point).Set(point)
	for _, scalarByte := range ed25519Order {
		for bit := 0; bit < 8; bit++ {
			if scalarByte&(1<<bit) != 0 {
				result.Add(result, addend)
			}
			addend.Double(addend)
		}
	}
	if result.Equal(identity) != 1 {
		return fmt.Errorf("Edwards25519 point is not prime order")
	}
	return nil
}

// KeyIDFromSPKI returns the Makoto v0.2 key ID for canonical Ed25519 SPKI.
func KeyIDFromSPKI(spki []byte) (string, error) {
	if _, err := ParsePublicSPKI(spki); err != nil {
		return "", err
	}
	sum := sha256.Sum256(spki)
	return "sha256:" + hex.EncodeToString(sum[:]), nil
}

// SignEnvelope signs exact payload bytes with distinct Ed25519 keys. Signature
// order is canonical by key ID.
func SignEnvelope(
	payloadType string,
	payload []byte,
	privateKeys ...ed25519.PrivateKey,
) (*Envelope, error) {
	if err := ValidatePayloadType(payloadType); err != nil {
		return nil, err
	}
	if len(privateKeys) == 0 {
		return nil, fmt.Errorf("at least one signing key is required")
	}
	message := PAE(payloadType, payload)
	signatures := make([]Signature, 0, len(privateKeys))
	seen := make(map[string]struct{}, len(privateKeys))
	for _, privateKey := range privateKeys {
		if len(privateKey) != ed25519.PrivateKeySize {
			return nil, fmt.Errorf("Ed25519 private key must be exactly 64 bytes")
		}
		publicKey := privateKey.Public().(ed25519.PublicKey)
		spki, err := PublicSPKI(publicKey)
		if err != nil {
			return nil, err
		}
		keyID, err := KeyIDFromSPKI(spki)
		if err != nil {
			return nil, err
		}
		if _, duplicate := seen[keyID]; duplicate {
			return nil, fmt.Errorf("signing key IDs must be unique")
		}
		seen[keyID] = struct{}{}
		signatures = append(signatures, Signature{
			KeyID: keyID,
			Sig:   base64.StdEncoding.EncodeToString(ed25519.Sign(privateKey, message)),
		})
	}
	sort.Slice(signatures, func(i, j int) bool {
		return signatures[i].KeyID < signatures[j].KeyID
	})
	return &Envelope{
		PayloadType: payloadType,
		Payload:     base64.StdEncoding.EncodeToString(payload),
		Signatures:  signatures,
	}, nil
}

func scalarIsCanonical(scalar []byte) bool {
	for index := len(ed25519Order) - 1; index >= 0; index-- {
		if scalar[index] < ed25519Order[index] {
			return true
		}
		if scalar[index] > ed25519Order[index] {
			return false
		}
	}
	return false
}

// VerifyEnvelopeSignature verifies one selected Ed25519 signature over the
// envelope's exact payload bytes.
func VerifyEnvelopeSignature(
	envelope *Envelope,
	publicSPKI []byte,
	signatureIndex int,
) error {
	if envelope == nil {
		return fmt.Errorf("envelope is nil")
	}
	if err := ValidatePayloadType(envelope.PayloadType); err != nil {
		return err
	}
	payload, err := canonicalBase64Decode(envelope.Payload, -1)
	if err != nil {
		return fmt.Errorf("decode payload: %w", err)
	}
	if signatureIndex < 0 || signatureIndex >= len(envelope.Signatures) {
		return fmt.Errorf("signature index is unavailable")
	}
	seen := make(map[string]struct{}, len(envelope.Signatures))
	decodedSignatures := make([][]byte, len(envelope.Signatures))
	for index, signature := range envelope.Signatures {
		if len(signature.UnknownFields) != 0 {
			return fmt.Errorf("signature entry is malformed")
		}
		if !keyIDPattern.MatchString(signature.KeyID) {
			return fmt.Errorf("signature key ID is malformed")
		}
		if _, duplicate := seen[signature.KeyID]; duplicate {
			return fmt.Errorf("signature key IDs must be unique")
		}
		seen[signature.KeyID] = struct{}{}
		decoded, decodeErr := canonicalBase64Decode(signature.Sig, ed25519.SignatureSize)
		if decodeErr != nil {
			return fmt.Errorf("decode signature: %w", decodeErr)
		}
		decodedSignatures[index] = decoded
	}
	publicKey, err := ParsePublicSPKI(publicSPKI)
	if err != nil {
		return err
	}
	expectedKeyID, err := KeyIDFromSPKI(publicSPKI)
	if err != nil {
		return err
	}
	entry := envelope.Signatures[signatureIndex]
	if entry.KeyID != expectedKeyID {
		return fmt.Errorf("signature key ID does not match the supplied public key")
	}
	signature := decodedSignatures[signatureIndex]
	if err := validatePrimeOrderPoint(signature[:32]); err != nil {
		return fmt.Errorf("invalid Ed25519 signature point: %w", err)
	}
	if !scalarIsCanonical(signature[32:]) {
		return fmt.Errorf("noncanonical Ed25519 scalar")
	}
	if !ed25519.Verify(publicKey, PAE(envelope.PayloadType, payload), signature) {
		return fmt.Errorf("Ed25519 signature verification failed")
	}
	return nil
}

// PayloadBytes decodes the envelope's canonical base64 payload.
func (e Envelope) PayloadBytes() ([]byte, error) {
	return canonicalBase64Decode(e.Payload, -1)
}
